### ISSUE TITLE
Requirements Doc

### ISSUE THREAD

**Issue by abhijithneilabraham** [vitalops/datatune#48]
# Datatune Requirements: Intelligent Operation Selection

## Overview

Transform Datatune from manual operation selection to intelligent automatic selection using LLMs. Currently supports map/filter - expanding to full data operation suite with multiple LLM deployment strategies.

## Core Problem

Users shouldn't need to know whether they want map, filter, or reduce. They should just describe what they want: "get all users over 25" or "calculate average salary by department" and Datatune figures out the operations.

### Operation Selection Engine

**Intent Parser**: Understand user intent and classify into operation types through semantic analysis
- Filtering intent: "show me users over 25", "I need customers from California" → filter 
- Transformation intent: "convert all prices to euros", "make names uppercase" → map
- Aggregation intent: "what's the total revenue", "average salary per department" → reduce
- Ordering intent: "rank by performance", "show highest to lowest" → sort
- Grouping intent: "break down by category", "organize users by region" → groupby

**Prompt → Operation Mapping**:
```python
datatune.query(df, "get users over 25") → internally selects filter operation
datatune.query(df, "convert prices to USD") → internally selects map operation
datatune.query(df, "total sales by region") → internally selects groupby + reduce
datatune.query(df, "top 10 customers by revenue") → internally selects sort + limit
```

**Multi-Operation Detection**: Handle complex prompts requiring chains
```python
datatune.query(df, "get top 5 customers by revenue") 
# → internally: sort by revenue desc → take first 5

datatune.query(df, "average age of users in each city")
# → internally: group by city → calculate mean of age
```

### LLM Integration Strategy

**Three-Tier Approach**:

1. **Fast Semantic Classifier** (Primary)
   - Small sentence embedding model for intent classification
   - <100ms response time
   - Handles 80% of use cases through semantic similarity
   - Fallback to full LLM for complex cases

2. **Full LLM** (Secondary)  
   - For complex/ambiguous prompts
   - Parameter extraction and validation
   - Multi-step operation planning

3. **Hybrid Approach**
   - Semantic classifier for initial intent detection
   - LLM for confidence < threshold or complex semantic understanding
   - Learn from LLM responses to improve semantic model

### 1. Bring Your Own LLM (BYOL)
Users provide their own API keys/endpoints
```python
datatune.configure(llm_provider="openai", api_key="...")
```

### 2. Self-Hosted Infrastructure
Deploy LLM in user's infrastructure
```python
datatune.configure(llm_endpoint="http://internal-llm:8080")
```

### 3. Datatune Credits System
Hosted service, users buy credits
```python
datatune.configure(credits_token="dt_abc123")
```

**Credits Pricing Strategy**:
- Simple operations (filter/map): 0.1 credits
- Complex operations (multi-step): 0.5 credits  
- Custom model training: 10 credits/hour
- Free tier: 100 credits/month

### 4. Offline Mode
Pre-trained lightweight model bundled with package
- No internet required
- Limited to common operations
- Fallback when other options unavailable

### Operation Selection Algorithm

**Step 1: Semantic Classification**
```python
# Internal operation selection process (user doesn't see this)
def select_operation(prompt, data_schema):
    # Analyze user intent semantically
    intent_classifier = load_semantic_model()
    operation_intent = intent_classifier.classify(prompt)
    
    # Return operation plan
    return {
        'primary_operation': 'filter',
        'confidence': 0.95,
        'execution_plan': ['filter_by_age', 'filter_by_location']
    }
```

**Step 2: Parameter Extraction & Validation**
```python
# Extract specific conditions from natural language
def extract_parameters(prompt, selected_operation, data_schema):
    # Use LLM to understand what user wants to filter/transform
    llm_response = llm.extract_conditions(
        prompt="get users over 25 in California",
        operation="filter",
        available_columns=data_schema.columns
    )
    
    return {
        'operation': 'filter',
        'conditions': {
            'age': {'operator': '>', 'value': 25},
            'state': {'operator': '==', 'value': 'California'}
        }
    }
```

**Step 3: Execution**
```python
# User-facing API remains simple
result = datatune.query(df, "get users over 25 in California")

# Internally, Datatune executes the planned operations
# User doesn't need to know about pandas/numpy specifics
```

**Operation Validation**:
- Dry-run operations on sample data
- Catch errors before full execution
- Suggest corrections for failed operations

### 1.  Explainable Operations

User can inspect the operation plan
```
print(result.operation_plan)
# {'operation': 'filter', 'column': 'customer_value', 'threshold': 10000}
```

### 2. Operation Composition
Handle complex multi-step operations:
```python
result = datatune.query(df, "top 10 customers by revenue in each region")

# Datatune automatically plans and executes:
# 1. Group by region
# 2. Sort by revenue within each group  
# 3. Take top 10 from each group

print(result.operation_plan)
# {'operations': ['group_by_region', 'sort_by_revenue_desc', 'top_n_per_group'], 'n': 10}
```

### Training Data for Operation Selection

**Semantic Data Generation**:
- Generate 10k+ natural language prompt-operation pairs
- Cover diverse ways of expressing the same intent
- Include colloquial, technical, and domain-specific language variations

**Real User Data Collection**:
- Opt-in telemetry for operation success/failure
- A/B testing for operation selection accuracy
- Feedback loop for continuous improvement

### Model Training Pipeline

**Fine-tuning Strategy**:
- Start with pre-trained sentence embedding model
- Fine-tune on data operation intent classification
- Train on semantic similarity rather than exact matching
- Target: <50MB model size for embeddings + classifier

### Technical Metrics
- Operation selection accuracy: >92%
- Response time: <200ms (local), <2s (LLM)
- Cost per operation: <$0.001
- Cache hit rate: >70%

### User Experience Metrics
- Reduced time to insight: 50% improvement
- User satisfaction: >4.5/5
- Adoption rate: 10x increase in active users

### Phase 1: Core Engine 
- Semantic intent classification using sentence embeddings
- Natural language understanding for basic operations
- Local execution engine

### Phase 2: LLM Integration
- Provider abstraction layer
- Parameter extraction
- Complex operation detection

### Phase 3: Smart Features
- Multi-operation chaining
- Error recovery

### Phase 4: Production Ready 
- Credits system
- Monitoring and analytics
- Enterprise features

### Technical Risks
- **Operation Misclassification**: Confidence thresholds + user feedback
- **LLM Provider Failures**: Multi-provider support + local fallback

## Open Questions

1. **Custom Operation Support**: How to handle domain-specific operations?
2. **Multi-language Support**: Beyond English prompts?
3. **Streaming Operations**: Real-time data processing?
4. **Model Personalization**: Per-user vs per-organization models?
5. **Integration Complexity**: How to handle complex data pipelines?

## Next Steps

1. Build semantic intent classifier for common data operations
2. Design datatune.query() API interface and response format
3. Implement LLM integration for parameter extraction
4. Create operation planning and execution engine
5. Develop confidence scoring and clarification system
